File: src/common/git.js
var exec = require('child_process').exec;
var q = require("q");  // npm install q
var fs = require('fs')
var path = require('path');

module.exports.cloneRepo = function(gitRemote, destination){
    return execGitCmd(`git clone --depth 1 ${gitRemote} ${destination}`)
}

module.exports.validateBundle = function(repoPath, bundlePath){
    return execGitCmd(`git bundle verify ${bundlePath}`, repoPath)
}

module.exports.bundleBranchName = function(bundlePath){
    return execGitCmd(`git bundle list-heads ${bundlePath} | awk 'NR==1{ print $2; }'`)
}

module.exports.fetchBundleCommits = function(repoPath, bundlePath, bundleBranchName, localBranchName){
    return execGitCmd(`git fetch ${bundlePath} ${bundleBranchName}:${localBranchName}`, repoPath)
}

module.exports.checkoutBranch = function checkoutBranch(repoPath, branchName){
    return execGitCmd(`git checkout ${branchName}`, repoPath)
}

module.exports.squashCommits = function(repoPath, destBranchName, squashedBranchName, bundleBranchName){
    //create a new squashedBranch, which is based off the destBranch.
    return execGitCmd(`git checkout -b ${squashedBranchName} ${destBranchName}`, repoPath)
        .then(function(){
            //now lets merge (and squash) the commits from the git bundle.
            return execGitCmd(`git merge --squash ${bundleBranchName}`, repoPath);
        })
        .then(function(){
            //merge --squash doesn't actually create a commmit, so we need to do that here.
            // we also need to set the commiter and author
            process.env.GIT_COMMITTER_NAME = "ghost";
            process.env.GIT_COMMITTER_EMAIL = "ghost@users.noreply.github.com";
            process.env.GIT_AUTHOR_NAME = "ghost";
            process.env.GIT_AUTHOR_EMAIL = "ghost@users.noreply.github.com";

            return execGitCmd(`git commit -am "anonymous commit"`, repoPath, process.env)
        })


    return execGitCmd(`git push origin ${branchName}`, repoPath)
}

module.exports.pushRepo = function(repoPath, branchName){
    return execGitCmd(`git push origin ${branchName}`, repoPath)
}

var GitPath = '/tmp/git.gitmask';
var gitExpand = false;

function execGitCmd(cmd, cwd, env){
    if (gitExpand && fs.existsSync(GitPath)) {
        // Use the existing one as it takes much time to expand git everytime.
        return execGitCmdImpl(cmd, cwd, env);
    }
    return require('lambda-git')({targetDirectory: GitPath}).then(function(){
        gitExpand = true;
        return execGitCmdImpl(cmd, cwd, env);
    });
}

function execGitCmdImpl(cmd, cwd, env){
    var opts = {}
    if(cwd){
        opts.cwd = cwd;
    }
    if(env){
        opts.env = env
    }
    else{
        opts.env = process.env
    }

    var deferred = q.defer();
    exec(cmd, opts, function(err, stdout, stderr) {
        console.log(" >> STDERR:", stderr)
        if (err) return deferred.reject(err);
        return deferred.resolve(stdout)
    });
    return deferred.promise;
}

File: gitmask/lib/common/git.py
from dulwich.porcelain import (clone, pull, update_head, branch_list, active_branch)
import subprocess
import os

def repo_clone(git_remote, repo_path, is_bare=False):
    # '--depth', '1',

    cmd = ['git', 'clone']
    if is_bare:
        cmd.append('--bare')
    cmd.append(git_remote)
    cmd.append(repo_path)

    process = subprocess.check_call(cmd)


def repo_fetch(repo_path):
    process_fetch = subprocess.check_call(['git', 'fetch', '--all'], cwd=repo_path)

def repo_checkout(repo_path, branch_name):
    process = subprocess.check_call(['git', 'checkout', branch_name], cwd=repo_path)

def repo_squash_commits(repo_path, dest_branch, squashed_branch, packfile_branch):

    process = subprocess.check_call(['git', 'checkout', '-b', squashed_branch, dest_branch], cwd=repo_path)

    process_merge = subprocess.check_call(['git', 'merge', '--squash', packfile_branch], cwd=repo_path)

    cust_env = os.environ.copy()
    cust_env['GIT_COMMITTER_NAME'] = 'ghost'
    cust_env['GIT_COMMITTER_EMAIL'] = 'ghost@users.noreply.github.com'
    cust_env['GIT_AUTHOR_NAME'] = 'ghost'
    cust_env['GIT_AUTHOR_EMAIL'] = 'ghost@users.noreply.github.com'

    # make a squashed commit on the squashed_branch
    process_add = subprocess.check_call(['git', 'commit', '-am', 'gitmask.com anonymous commit'], cwd=repo_path, env=cust_env)


def repo_push(repo_path, remote_name, branch_name):
    process = subprocess.check_call(['git', 'push', remote_name, branch_name], cwd=repo_path)

def repo_remote_add(repo_path, remote_name, remote_url):
    process = subprocess.check_call(['git', 'remote', 'add', remote_name, remote_url], cwd=repo_path)


#
#
# def clone_repo(gitRemote, destination, branch, depth=1):
#     branch_ref = bytes('refs/heads/{0}'.format(branch), 'utf-8')
#
#     local_repo = clone(gitRemote, destination, bare=False)
#     pull(destination, refspecs=[b'HEAD', branch_ref])
#
#     print("=====AFTER=====>", branch_list(destination))
#     update_head(destination, branch_ref, False)
#     local_repo.reset_index(local_repo[branch_ref].tree)
#     local_repo.refs.set_symbolic_ref(b'HEAD', branch_ref)
#     print("ACTIVE BRANCH====>", active_branch(destination))

# def checkout_branch(repoPath, branchName):
#     return execGitCmd(`git checkout ${branchName}`, repoPath)

